File: app/src/main/java/com/airbnb/net/AirbnbAdapter.java
package com.airbnb.net;

import java.util.List;

import retrofit.MockRestAdapter;
import retrofit.RestAdapter;
import rx.Observable;

public class AirbnbAdapter implements AirbnbService{
    private static final boolean USE_MOCK_ADAPTER = true;


    private static AirbnbAdapter sInstance;

    public static synchronized AirbnbAdapter getInstance() {
        if (sInstance == null) {
            sInstance = new AirbnbAdapter();
        }

        return sInstance;
    }

    private AirbnbService mService;

    private AirbnbAdapter() {
        RestAdapter adapter = new RestAdapter.Builder()
                .setEndpoint("http://www.airbnb.com/api/v1")
                .build();

        if (USE_MOCK_ADAPTER) {
            MockRestAdapter mockAdapter = MockRestAdapter.from(adapter);
            mockAdapter.setDelay(250);
            mockAdapter.setVariancePercentage(50);
            mockAdapter.setErrorPercentage(0);
            mService = mockAdapter.create(AirbnbService.class, new MockAirbnbService());
        } else {
            mService = adapter.create(AirbnbService.class);
        }
    }

    @Override
    public Observable<List<ListItem>> getSearchTabItems() {
        return mService.getSearchTabItems();
    }

    @Override
    public Observable<List<HeroItem>> getWishList() {
        return mService.getWishList();
    }
}

File: app/src/main/java/com/airbnb/net/AirbnbAdapter.kt
package com.airbnb.net

import com.airbnb.Message

import retrofit.MockRestAdapter
import retrofit.RestAdapter
import retrofit.http.Path
import rx.Observable
import kotlin.properties.Delegates

public object AirbnbAdapter : AirbnbService {
    private val USE_MOCK_ADAPTER = true

    private var mService: AirbnbService

    init {
        val adapter = RestAdapter.Builder()
                .setEndpoint("http://www.airbnb.com/api/v1")
                .build()

        if (USE_MOCK_ADAPTER) {
            val mockAdapter = MockRestAdapter.from(adapter)
            mockAdapter.setDelay(250)
            mockAdapter.setVariancePercentage(50)
            mockAdapter.setErrorPercentage(0)
            mService = mockAdapter.create(javaClass<AirbnbService>(), MockAirbnbService())
        } else {
            mService = adapter.create(javaClass<AirbnbService>())
        }
    }

    override fun getSearchTabItems(): Observable<List<ListItem>> {
        return mService.getSearchTabItems()
    }

    override fun getWishList(): Observable<List<HeroItem>> {
        return mService.getWishList()
    }

    override fun getMessages(): Observable<List<Message>> {
        return mService.getMessages()
    }

    override fun getListing(Path("id") id: Long): Observable<Listing> {
        return mService.getListing(id)
    }
}
